Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,251 @@
+The view-distance config value will determine the global distance (and set maximum) of normal view distance interactions with entities, mob-spawning and tracking.
+All chunks with player tickets within this range will also be served by the server, allowing shared view distance from other players without impacting performance.
+
+diff --git a/src/main/java/net/minecraft/server/ChunkMapDistance.java b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+index 63a68872..ed4d6d09 100644
+--- a/src/main/java/net/minecraft/server/ChunkMapDistance.java
++++ b/src/main/java/net/minecraft/server/ChunkMapDistance.java
+@@ -43,6 +43,16 @@ public abstract class ChunkMapDistance {
+     private final LongSet l = new LongOpenHashSet();
+     private final Executor m;
+     private long currentTick;
++    //Paper start ticket level pp-view-dist
++    protected final it.unimi.dsi.fastutil.ints.Int2ObjectMap<Square> squares = new it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap<>();
++    protected it.unimi.dsi.fastutil.longs.Long2ByteMap update = new it.unimi.dsi.fastutil.longs.Long2ByteOpenHashMap();
++    class Square {
++        public int x, y, u, v, id;
++        public Square(int x1, int y1, int u1, int v1, int id1) {
++            x = x1; y = y1; u = u1; v = v1; id = id1;
++        }
++    }
++    //Paper end
+ 
+     protected ChunkMapDistance(Executor executor, Executor executor1) {
+         executor1.getClass();
+@@ -228,12 +238,95 @@ public abstract class ChunkMapDistance {
+ 
+     }
+ 
+-    public void a(SectionPosition sectionposition, EntityPlayer entityplayer) {
++    //Paper start ticket level pp-view-dist
++    public void a(SectionPosition sectionposition, EntityPlayer pln) {
+         long i = sectionposition.u().pair();
+ 
+         ((ObjectSet) this.c.computeIfAbsent(i, (j) -> {
+             return new ObjectOpenHashSet();
+-        })).add(entityplayer);
++        })).add(pln);
++
++        int cx = ChunkCoordIntPair.getX(i);
++        int cz = ChunkCoordIntPair.getZ(i);
++
++        int pvd = pln.getViewDistance() - 2;
++
++        Square vdsqr = new Square(cx - pvd, cz - pvd, cx + pvd, cz + pvd, 0);
++        Square sto = this.squares.get(pln.getUniqueID().hashCode());
++        this.squares.put(pln.getUniqueID().hashCode(), new Square(cx - pvd, cz - pvd, cx + pvd, cz + pvd, 0));
++
++        if(sto == null) sto = vdsqr;
++        Square oldsqr = new Square(sto.x, sto.y, sto.u, sto.v, sto.id);
++
++        it.unimi.dsi.fastutil.ints.IntSortedSet row = new it.unimi.dsi.fastutil.ints.IntRBTreeSet(it.unimi.dsi.fastutil.ints.IntComparators.NATURAL_COMPARATOR);
++        it.unimi.dsi.fastutil.ints.IntSortedSet col = new it.unimi.dsi.fastutil.ints.IntRBTreeSet(it.unimi.dsi.fastutil.ints.IntComparators.NATURAL_COMPARATOR);
++        it.unimi.dsi.fastutil.objects.ObjectList<Square> toproc = new it.unimi.dsi.fastutil.objects.ObjectArrayList<>();
++
++        if(oldsqr != null) {
++            vdsqr.id = 1; oldsqr.id = -1;
```

Comment:
if we're using the id as a state we need to extract them to static final constants, appropriately named, so that when we go back and look at the pr we know what they're supposed to do

Reply:
this was previously a boolean, only changed to a byte since booleans default null as false. I'll change the var to `add` or something similar.